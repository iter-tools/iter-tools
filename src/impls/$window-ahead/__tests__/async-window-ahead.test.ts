/**
 * @generated-from ./$window-ahead.test.ts
 * This file is autogenerated from a template. Please do not edit it directly.
 * To rebuild it from its template use the command
 * > npm run generate
 * More information can be found in CONTRIBUTING.md
 */

import { asyncWindowAhead } from 'iter-tools-es';
import { asyncWrap, asyncUnwrapDeep } from '../../../test/async-helpers.js';

describe('asyncWindowAhead', () => {
  describe('when source is empty', () => {
    it('yields no windows', async () => {
      expect(await asyncUnwrapDeep(asyncWindowAhead({ filler: 0 }, 3, null))).toEqual([]);
      expect(await asyncUnwrapDeep(asyncWindowAhead({ filler: 0 }, 3, undefined))).toEqual([]);
      expect(await asyncUnwrapDeep(asyncWindowAhead({ filler: 0 }, 3, asyncWrap([])))).toEqual([]);
    });
  });

  describe('when size(source) < size', () => {
    it('yields only partial windows', async () => {
      expect(await asyncUnwrapDeep(asyncWindowAhead({ filler: 0 }, 3, asyncWrap([1, 2])))).toEqual([
        [1, 2, 0],
        [2, 0, 0],
      ]);
    });
  });

  describe('when size(source) === size', () => {
    it('yields one full window, then partial windows', async () => {
      expect(
        await asyncUnwrapDeep(asyncWindowAhead({ filler: 0 }, 3, asyncWrap([1, 2, 3]))),
      ).toEqual([
        [1, 2, 3],
        [2, 3, 0],
        [3, 0, 0],
      ]);
    });
  });

  describe('when size(source) > size', () => {
    it('yields size(source)-size full windows, then partial windows', async () => {
      expect(
        await asyncUnwrapDeep(asyncWindowAhead({ filler: 0 }, 2, asyncWrap([1, 2, 3]))),
      ).toEqual([
        [1, 2],
        [2, 3],
        [3, 0],
      ]);
    });
  });

  describe('when useFiller is false', () => {
    it('yields shorter windows instead of windows with filler', async () => {
      expect(
        await asyncUnwrapDeep(asyncWindowAhead({ useFiller: false }, 3, asyncWrap([1, 2, 3]))),
      ).toEqual([[1, 2, 3], [2, 3], [3]]);
    });
  });

  it('has a default filler of undefined', async () => {
    expect(await asyncUnwrapDeep(asyncWindowAhead(2, asyncWrap([1])))).toEqual([[1, undefined]]);
  });

  describe('when arguments are invalid', () => {
    it('throws', async () => {
      expect(() =>
        asyncWindowAhead('foo' as any, asyncWrap([1, 2, 3])),
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
