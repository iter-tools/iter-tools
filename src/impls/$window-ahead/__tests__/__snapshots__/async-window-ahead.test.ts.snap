// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asyncWindowAhead when arguments are invalid throws 1`] = `"asyncWindowAhead must be passed a numeric size"`;
